services/gitlab/gitlab-tag.spec.js

```diff
@@ -0,0 +1,47 @@
+import { expect } from 'chai'
+import nock from 'nock'
+import { cleanUpNockAfterEach, defaultContext } from '../test-helpers.js'
+import GitLabTag from './gitlab-tag.service.js'
+
+describe('GitLabTag', function () {
+  describe('auth', function () {
+    cleanUpNockAfterEach()
+
+    const fakeToken = 'abc123'
+    const config = {
+      public: {
+        services: {
+          gitlab: {
+            authorizedOrigins: ['https://gitlab.com'],
+          },
+        },
+      },
+      private: {
+        gitlab_token: fakeToken,
+      },
+    }
+
+    it('sends the auth information as configured', async function () {
```

Comment:
My reading of the code here is that we'll send our token to gitlab.com (or whatever is in `GITLAB_ORIGINS`), but you can still specify a custom `gitlab_url` (that is not in `GITLAB_ORIGINS`) and it will work (e.g: https://shields-staging-pr-6988.herokuapp.com/gitlab/tag/GNOME/pango?gitlab_url=https%3A%2F%2Fgitlab.gnome.org does render a badge) but it just won't send the token for those requests. Two follow-ups for that:
1. Can we add a second test case here that explicitly verifies our token is not sent when we specify a custom `gitlab_url` (and the url is not in `GITLAB_ORIGINS`)
2. Do we think the anonymous rate limit becomes a problem for alternate gitlab installs? Presumably self-hosters can configure this differently to the centralised service.. Do you think it is necessary to allow users to also pass a `&token=` param? Maybe we can just ignore this to start with it and deal with if it becomes a problem. I'm just trying to think ahead a bit..

Reply:
> Can we add a second test case here that explicitly verifies our token is not sent when we specify a custom gitlab_url (and the url is not in GITLAB_ORIGINS)

I pulled in this test case because we have one such case for all other authenticated services. I guess I'm not strongly opposed but we don't have it for any of the similar services which I imagine is because the negative case is sufficiently covered in the auth helper tests, such as:

https://github.com/badges/shields/blob/f4bddb9964247a8d83803b3a59c13454e33c9cc6/core/base-service/auth-helper.spec.js#L344-L363

> Do we think the anonymous rate limit becomes a problem for alternate gitlab installs? Presumably self-hosters can configure this differently to the centralised service.. Do you think it is necessary to allow users to also pass a &token= param?

My reading of the GitLab docs on throttling is that it is IP based, so I don't think allowing users to provide a token nor even a hypothetical GitHub-like token pooling solution would be of any help for us on the rate limiting front.

GitLab does support scoped access tokens, including the ability to scope in a strictly read-only way. As such I think allowing users to provide their own token could help with cases of badges for private projects via Shields.io without having to self-host the badge server. However, I suspect that will pose a technical problem with our current auth injection mechanisms, since we'd need to pivot to a "don't worry about the authorized origins if it's a user provided token, but still enforce authorized origins with the configured token if one wasn't provided". Presumably a problem that can be overcome, but probably best done in a separate dedicated thread.

I don't think we'll have the ability to readily support authenticated requests for alternate instances without another type of non-trivial change to our auth story, as we'd need the ability to define multiple mappings of an auth mechanism to a set of authorized origins, whereas today it's strictly a 1 secret/token/creds to many origins.

Reply:
> we don't have it for any of the similar services which I imagine is because the negative case is sufficiently covered in the auth helper tests

OK yeah I think its covered by the tests for the base class.

> I don't think we'll have the ability to readily support authenticated requests for alternate instances without another type of non-trivial change to our auth story, as we'd need the ability to define multiple mappings of an auth mechanism to a set of authorized origins, whereas today it's strictly a 1 secret/token/creds to many origins.

Useful to have thought it through a bit, but I think we can just ignore it for now and cross this bridge if it ever becomes an issue. I think the anon rate limit should cover most custom server cases anyway.